### PR TITLE
base64 encoded credentials string for remote_storage: gcp

### LINKED ISSUE #498: allow to GCP JSON access credentials Base64 encoded
Hello, I am trying to create backup to remote google storage bucket.
My config is
```
  - name: REMOTE_STORAGE
     value: "gcs"
  - name: GCS_BUCKET
     value: "bucket-name"
  - name: GCS_PATH
     value: "clickhouse"
  - name: GCS_DEBUG
     value: "true"
  - name: GCS_CREDENTIALS_JSON
     value: "base64 encoded string that contains credentials json data"
```
however, I am getting error 
`error	can\'t connect to GCS: googleHTTPTransport.NewClient error: invalid character \'e\' looking for beginning of value`
I believe that I passing wrong values in "GCS_CREDENTIALS_JSON". How can I fix it ?

### PR DESCRIPTION
**Pull request by navi86**
this should allow to pass encoded string to "WithCredentialsJSON".

fix https://github.com/AlexAkulov/clickhouse-backup/issues/498

### CODE REVIEW

File: pkg/storage/gcs.go [hunk 1]
```diff
@@ -64,7 +66,8 @@ func (gcs *GCS) Connect() error {
 		clientOptions = append([]option.ClientOption{option.WithoutAuthentication()}, clientOptions...)
 		clientOptions = append(clientOptions, option.WithEndpoint(endpoint))
 	} else if gcs.Config.CredentialsJSON != "" {
-		clientOptions = append(clientOptions, option.WithCredentialsJSON([]byte(gcs.Config.CredentialsJSON)))
+		d, _ := base64.StdEncoding.DecodeString(gcs.Config.CredentialsJSON)
```
Comment (Slach) [inline pkg/storage/gcs.go:71]: Sorry, these changes will break compatibility with current behavior

Could you implement it as a separate config option?
Like `gcs.Config.CredentialsJSONEncoded`

**Review by Slach — CHANGES_REQUESTED** [review]

File: pkg/storage/gcs.go (hunk 1 above)
Comment (Slach) [reply pkg/storage/gcs.go:71]: Moreover, could you add tests/integration/integration_test.go test case to cover these changes?

Comment (navi86) [reply pkg/storage/gcs.go:71]: @Slach thanks for reviewing, I don't see any integration test for CredentialsJSON, I only see its for "WithCredentialsFile". Could you share the link to it ? Moreover, how can I run integration tests locally ?

Comment (Slach) [reply pkg/storage/gcs.go:71]: unfortunately, no one test case for `CredentialsJSON` in `tests/integration/` only in `tests/testflows` in python

look https://github.com/AlexAkulov/clickhouse-backup/search?q=QA_GCS_CRED_JSON for details

Comment (navi86) [reply pkg/storage/gcs.go:71]: @Slach I am not bale to test it properly
first of all, I have not related issue
https://github.com/golang/go/issues/54291
secondly, if I run without options "-gcflags "all=-N -l" it runs successfully, then, if I run test
`python3 ./test/testflows/clickhouse_backup/regression.py --only ".test/testflows/clickhouse_backup/tests/cloud_storage.py"`
I got 
`docker.errors.DockerException: Error while fetching server API version: ('Connection aborted.', FileNotFoundError(2, 'No such file or directory'))`
could you helm how I can configure env for testing on MacOS ?

Comment (Slach) [reply pkg/storage/gcs.go:71]: did you read 
https://github.com/AlexAkulov/clickhouse-backup/blob/master/test/testflows/README.md ?

Comment (navi86) [reply pkg/storage/gcs.go:71]: @Slach 
yes, but it's not working on my pc.

File: test/testflows/clickhouse_backup/tests/cloud_storage.py [hunk 2]
```diff
@@ -194,8 +194,8 @@ def s3_aws(self):
 def gcs(self):
     """Test that an existing backup can be uploaded to GCS bucket.
     """
-    if not os.environ.get('QA_GCS_CRED_JSON'):
-        skip("QA_GCS_CRED_JSON environment not found, test skipped")
+    if not (os.environ.get('QA_GCS_CRED_JSON') or os.environ.get('QA_GCS_CRED_JSON_ENCODED')):
+        skip("QA_GCS_CRED_JSON/QA_GCS_CRED_JSON_ENCODED environment not found, test skipped")
 
     test_storage_outline(storage_type="gcs")
```
Comment (Slach) [inline test/testflows/clickhouse_backup/tests/cloud_storage.py:200]: did you look inside test_storage_outline? you shall handle `QA_GCS_CRED_JSON_ENCODED`

**Review by Slach — CHANGES_REQUESTED** [review]

File: test/testflows/clickhouse_backup/tests/cloud_storage.py (hunk 2 above)
Comment (navi86) [reply test/testflows/clickhouse_backup/tests/cloud_storage.py:200]: yes, I've updated code

Comment (Slach) [reply test/testflows/clickhouse_backup/tests/cloud_storage.py:200]: but test still not coverage `QA_GCS_CRED_JSON_ENCODED`

cause it not handle `inside test_storage_outline`

**Review by Slach — CHANGES_REQUESTED** [review]

**Comment by Slach** [comment]
@navi86 
looks fail https://github.com/AlexAkulov/clickhouse-backup/runs/7944825381?check_suite_focus=true#step:6:1465
to fix it
create properly `./test/testflows/.env` as described in `./test/testflows/README.md`
```
source ./tests/testflow/.env
rm -fv ./test/testflows/clickhouse_backup/tests/snapshots/cli.py.cli.snapshot
python3 ./tests/testflows/regression.py --only="/clickhouse backup/cli/default config"
```

**Comment by navi86** [comment]
@Slach 
thanks for info, I saw there is issue with tests,  I am not able to run it locally, because 
1. "python3 ./tests/testflows/regression.py --only="/clickhouse backup/cli/default config"" doesn't start time to time, for instance, I can run command without any progress and wait about 20 minutes so I had to cancel and run it again when it finally starts.
```
                23ms              [bash] bash#
              4m 32s         ⟥    Exception: termios.error: (4, 'Interrupted system call')
              4m 32s         ⟥⟤ Error executing command, /clickhouse backup/docker-compose cluster/I list environment variables to show their values/executing command, error
              4m 32s       ⟥⟤ Error I list environment variables to show their values, /clickhouse backup/docker-compose cluster/I list environment variables to show their values, error
              4m 32s     ⟥⟤ Error docker-compose cluster, /clickhouse backup/docker-compose cluster, error
              4m 32s   ⟥⟤ Error clickhouse backup, /clickhouse backup, error
                         termios.error: (4, 'Interrupted system call')

Failing

✘ [ Error ] /clickhouse backup
```
you can notice that it was running almost 5 mins without any progress
2. tests are failed because docker-compose is not able to start clickhouse container due to permissions error
`chown: changing ownership of '/var/log/clickhouse-server': Permission denied`
I don't know how to run this tests locally because it seems it doesn't work on my pc at all.

**Comment by Slach** [comment]
>> chown: changing ownership of '/var/log/clickhouse-server': Permission denied
> I don't know how to run this tests locally because it seems it doesn't work on my pc at all.

Which OS do you use for run docker?

**Comment by Slach** [comment]
image

**Comment by navi86** [comment]
> > > chown: changing ownership of '/var/log/clickhouse-server': Permission denied
> > > I don't know how to run this tests locally because it seems it doesn't work on my pc at all.
> 
> Which OS do you use for run docker?

MacOS and I am using lima instead Docker

**Comment by Slach** [comment]
did you install `docker` and `docker-compose` inside `lima` VM?

**Comment by navi86** [comment]
> did you install `docker` and `docker-compose` inside `lima` VM?

Yes, of course it's installed

**Review by Slach — CHANGES_REQUESTED** [review]
tesflows still failed
and https://github.com/AlexAkulov/clickhouse-backup/pull/499#discussion_r951008842 still not resolved

**Comment by navi86** [comment]
> tesflows still failed
> and [#499 (comment)](https://github.com/AlexAkulov/clickhouse-backup/pull/499#discussion_r951008842) still not resolved

oh, ok, I will check today later

**Comment by navi86** [comment]
@Slach I don't see any coverage inside test_storage_outline regarding of non encoded string:
https://github.com/AlexAkulov/clickhouse-backup/blob/master/test/testflows/clickhouse_backup/tests/cloud_storage.py#L91
I am not able to run tests because it simply doesn't work on mac. 
my changes is pretty small, could you run it and update file "cli.py.cli.snapshot" ?

**Pull request closed without merging**